### DISCUSSION TITLE
Is it possible to unlock password manager with pre hook?

### DISCUSSION THREAD

**Discussion by stinovlas** [twpayne/chezmoi#3750] [Q&A] [answered]
Hi all! I've been wondering whether I could leverage chezmoi hook system to unlock my bitwarden password manager before `apply` and then lock it again, after it's done. Unlocking bitwarden produces a key that has to be set to `BW_SESSION` environment variable. Is this somehow possible with chezmoi hooks?

It seems that chezmoi runs the hook commands in the subshell (which makes sense), but that makes it impossible to just export a variable inside the hook and affect the `apply` command. Any suggestions?

Also, it seems that the `post` hook is not applied when you quit the `apply` process or when an error occurs. I'd like to always call the post hook to lock the password manager.

I can of course write a wrapper script that will take care of all this, but I thought that using chezmoi hook system might be cleaner and more portable way.

**Comment by twpayne** [comment, marked as answer]
No, this is not possible on UNIX systems. See the discussion in https://github.com/twpayne/chezmoi/issues/3729.

**Reply by stinovlas** [reply]
I'm aware that it's not possible to export variable from the subprocess, I was just asking whether there is some registering mechanism with chezmoi (that could be done in number of ways, e.g. through a socket). Thanks for the link, I think I'll just create my own subcommand as descirbed in [docs](https://www.chezmoi.io/reference/plugins/) that will take care of this. I'm thinking about something like this:

```sh
#!/bin/sh
if ! bw login --check 2>/dev/null; then
  echo "You have to log in to bitwarden first!"
  exit 1
fi

if bw unlock --check 2>/dev/null
  export BW_SESSION=$(bw unlock --raw)
  should_lock=1
fi

chezmoi "$@"

if [ -n "$should_lock" ]; then
  bw lock
fi
```

**Reply by stinovlas** [reply]
Marking @twpayne reply as an answer, because it actually does answer the original question (the possibility using hooks to log in to bitwarden). Thank you for this great tool :-)!

**Comment by halostatue** [comment]
@twpayne We have received a half-dozen of these questions in the last 3–4 months. I wonder if it might not be worth calling something out either in the README and/or FAQ in chezmoi documentation. I’m definitely *not* able to tackle this right now, but I know that the various threads have exhaustive answers (including the strong suggestion that people consider `rbw` if they are bitwarden users).

**Reply by twpayne** [reply]
Yes, agreed. I have a list of user guide/FAQ entries that I want to write based on issues and discussions from the last few months.
